File: webapp-net/Core/Models/Entity/Teaser.cs
using Sdl.Web.Common.Models;
using System;
using System.Collections.Generic;
using System.ServiceModel.Syndication;
using Sdl.Web.Common.Configuration;

namespace Sdl.Web.Modules.Core.Models
{
    [SemanticEntity(EntityName = "Teaser", Prefix = "t", Vocab = CoreVocabulary)]
    [SemanticEntity(EntityName = "Image", Prefix = "i", Vocab = CoreVocabulary)]
    [SemanticEntity(EntityName = "Article", Prefix = "a", Vocab = CoreVocabulary)]
    [SemanticEntity(EntityName = "NewsArticle", Prefix = "na", Vocab = CoreVocabulary)]
    [SemanticEntity(EntityName = "Place", Prefix = "p", Vocab = CoreVocabulary)]
    [SemanticEntity(EntityName = "StandardMetadata", Prefix = "m", Vocab = CoreVocabulary)]
    [SemanticEntity(EntityName = "LinkedContent", Prefix = "c", Vocab = CoreVocabulary)]
    [Serializable]
    public class Teaser : EntityModel, ISyndicationFeedItemProvider
    {
        //A teaser can be mapped from an article or place, in which case the link should be to the item itself
        [SemanticProperty("na:_self")]
        [SemanticProperty("a:_self")]
        [SemanticProperty("p:_self")]
        [SemanticProperty("c:link")]

        public Link Link { get; set; }
        
        [SemanticProperty("na:name")]
        [SemanticProperty("a:name")]
        [SemanticProperty("headline")]
        [SemanticProperty("t:headline")]
        [SemanticProperty("a:headline")]
        [SemanticProperty("subheading")]
        [SemanticProperty("p:name")]
        [SemanticProperty("c:subheading")]
        public string Headline { get; set; }
        
        //A teaser can be mapped from an individual image, in which case the image property is set from the source entity itself
        [SemanticProperty("i:_self")]
        [SemanticProperty("a:image")]
        [SemanticProperty("na:image")]
        [SemanticProperty("c:media")]
        public MediaItem Media { get; set; }

        [SemanticProperty("na:introText")]
        [SemanticProperty("a:introText")]
        [SemanticProperty("content")]
        [SemanticProperty("description")]
        [SemanticProperty("m:description")]
        [SemanticProperty("c:content")]
        public RichText Text { get; set; }

        [SemanticProperty("dateCreated")]
        [SemanticProperty("m:dateCreated")]
        [SemanticProperty("c:date")]
        public DateTime? Date { get; set; }

        [SemanticProperty("p:location")]
        [SemanticProperty("c:location")]
        public Location Location { get; set; }

        //To store formatting options for the teaser (link style etc.)
        [SemanticProperty(IgnoreMapping=true)]
        private Dictionary<string, string> FormatOptions { get; set; }

        public string GetFormatOption(string key, string defaultValue = null)
        {
            if (FormatOptions != null && FormatOptions.ContainsKey(key))
            {
                return FormatOptions[key];
            }

            return defaultValue;
        }

        public void SetFormatOption(string key, string value)
        {
            if (FormatOptions == null)
            {
                FormatOptions = new Dictionary<string, string>();
            }
            if (!FormatOptions.ContainsKey(key))
            {
                FormatOptions.Add(key, value);
            }
            else
            {
                FormatOptions[key] = value;
            }
        }

        #region ISyndicationFeedItemProvider members
        /// <summary>
        /// Extracts syndication feed items.
        /// </summary>
        /// <param name="localization">The context <see cref="Localization"/>.</param>
        /// <returns>A single syndication feed item containing information extracted from this <see cref="Teaser"/>.</returns>
        public IEnumerable<SyndicationItem> ExtractSyndicationFeedItems(Localization localization)
        {
            Link link = Link;
            if (link == null && Media != null)
            {
                // If the Teaser doesn't have a Link, but does have Media, create a Link from its Media.
                link = new Link { Url = Media.Url };
            }
            return new[] { CreateSyndicationItem(Headline, Text, link, Date, localization) };
        }
        #endregion
    }
}

File: webapp-java/dxa-module-core/src/main/java/com/sdl/dxa/modules/core/model/entity/Teaser.java
package com.sdl.dxa.modules.core.model.entity;

import com.fasterxml.jackson.annotation.JsonProperty;
import com.google.common.collect.Lists;
import com.sdl.webapp.common.api.formatters.support.FeedItem;
import com.sdl.webapp.common.api.formatters.support.FeedItemsProvider;
import com.sdl.webapp.common.api.mapping.semantic.annotations.SemanticEntities;
import com.sdl.webapp.common.api.mapping.semantic.annotations.SemanticEntity;
import com.sdl.webapp.common.api.mapping.semantic.annotations.SemanticProperties;
import com.sdl.webapp.common.api.mapping.semantic.annotations.SemanticProperty;
import com.sdl.webapp.common.api.model.RichText;
import com.sdl.webapp.common.api.model.entity.AbstractEntityModel;
import com.sdl.webapp.common.api.model.entity.Link;
import com.sdl.webapp.common.api.model.entity.MediaItem;
import lombok.Data;
import lombok.EqualsAndHashCode;
import org.joda.time.DateTime;

import java.util.List;

import static com.sdl.webapp.common.api.mapping.semantic.config.SemanticVocabulary.SDL_CORE;

@SemanticEntities({
        @SemanticEntity(entityName = "Image", vocabulary = SDL_CORE, prefix = "i"),
        @SemanticEntity(entityName = "Article", vocabulary = SDL_CORE, prefix = "a"),
        @SemanticEntity(entityName = "NewsArticle", vocabulary = SDL_CORE, prefix = "na"),
        @SemanticEntity(entityName = "Place", vocabulary = SDL_CORE, prefix = "p"),
        @SemanticEntity(entityName = "LinkedContent", vocabulary = SDL_CORE, prefix = "c"),
        @SemanticEntity(entityName = "StandardMetadata", vocabulary = SDL_CORE, prefix = "m")
})
@Data
@EqualsAndHashCode(callSuper = true)
public class Teaser extends AbstractEntityModel implements FeedItemsProvider {

    @SemanticProperties({
            @SemanticProperty("na:_self"),
            @SemanticProperty("a:_self"),
            @SemanticProperty("p:_self"),
            @SemanticProperty("c:link")
    })
    @JsonProperty("Link")
    private Link link;

    @SemanticProperties({
            @SemanticProperty("m:name"),
            @SemanticProperty("na:name"),
            @SemanticProperty("a:name"),
            @SemanticProperty("headline"),
            @SemanticProperty("subheading"),
            @SemanticProperty("a:headline"),
            @SemanticProperty("p:name"),
            @SemanticProperty("c:headline"),
            @SemanticProperty("c:subheading")
    })
    @JsonProperty("Headline")
    private String headline;

    @SemanticProperties({
            @SemanticProperty("i:_self"),
            @SemanticProperty("a:image"),
            @SemanticProperty("c:media")
    })
    @JsonProperty("Media")
    private MediaItem media;

    @SemanticProperties({
            @SemanticProperty("m:introText"),
            @SemanticProperty("na:introText"),
            @SemanticProperty("a:introText"),
            @SemanticProperty("content"),
            @SemanticProperty("c:text"),
            @SemanticProperty("c:content"),
            @SemanticProperty("m:description"),
    })
    @JsonProperty("Text")
    private RichText text;

    @SemanticProperties({
            @SemanticProperty("c:date"),
            @SemanticProperty("m:dateCreated")
    })
    @JsonProperty("Date")
    private DateTime date;

    @SemanticProperties({
            @SemanticProperty("p:location"),
            @SemanticProperty("c:location")
    })
    @JsonProperty("Location")
    private Location location;

    public RichText getText() {
        return text != null ? text : new RichText("");
    }

    @Override
    public List<FeedItem> extractFeedItems() {
        return Lists.newArrayList(FeedItem.builder()
                .headline(headline)
                .summary(text)
                .link(getLinkForFeedItem())
                .date(date != null ? date.toDate() : null)
                .build());
    }

    private Link getLinkForFeedItem() {
        if (link != null) {
            return link;
        }
        if (media != null) {
            Link result = new Link();
            result.setUrl(media.getUrl());
            return result;
        }
        return null;
    }
}
